quota/etcd/storage/quota_storage.go

```diff
@@ -0,0 +1,387 @@
+// transaction.
+// By default, fn is only called for known, enabled configs. See forNamesMode for other behaviors.
+// Names are validated and de-duped automatically.
+func (qs *QuotaStorage) forNames(ctx context.Context, names []string, mode forNamesMode, fn func(concurrency.STM, string, *storagepb.Config) error) error {
+	for _, name := range names {
+		if !isNameValid(name) {
+			return fmt.Errorf("invalid name: %q", name)
+		}
+	}
+
+	_, err := concurrency.NewSTMSerializable(ctx, qs.Client, func(s concurrency.STM) error {
+		cfgs, err := getConfigs(s)
+		if err != nil {
+			return err
+		}
+
+		seenNames := make(map[string]bool)
+		for _, name := range names {
+			if seenNames[name] {
+				continue
+			}
+			seenNames[name] = true
+
+			emitted := false
+			for _, cfg := range cfgs.Configs {
+				if cfg.Name == name {
+					if cfg.State == storagepb.Config_ENABLED {
+						if err := fn(s, name, cfg); err != nil {
+							return err
+						}
+						emitted = true
+					}
+					break
+				}
+			}
+			if !emitted && mode == emitInfinite {
+				if err := fn(s, name, nil); err != nil {
+					return err
+				}
+			}
+		}
+		return nil
+	})
+	return err
+}
+
+func getConfigs(s concurrency.STM) (*storagepb.Configs, error) {
+	// TODO(codingllama): Consider watching configs instead of re-reading
+	cfgs := &storagepb.Configs{}
+	val := s.Get(configsKey)
+	if val == "" {
+		return cfgs, nil
+	}
+	if err := proto.Unmarshal([]byte(val), cfgs); err != nil {
+		return nil, fmt.Errorf("error unmarshaling %v: %v", configsKey, err)
+	}
+	return cfgs, nil
+}
+
+// modBucket adds "add" tokens to the specified quota. Add may be negative or zero.
+// Time-based quotas that are due replenishment will be replenished before the add operation. Quotas
+// that are above ceiling (eg, due to a config downgrade) will also be constrained to the
+// appropriate ceiling. As a consequence, calls with add = 0 are still useful for peeking and the
+// explained side-effects.
+// modBucket returns the current token count for cfg.
+func modBucket(s concurrency.STM, cfg *storagepb.Config, now time.Time, add int64) (int64, error) {
+	key := bucketKey(cfg)
+
+	val := s.Get(key)
+	bucket := &storagepb.Bucket{}
+	if err := proto.Unmarshal([]byte(val), bucket); err != nil {
+		return 0, fmt.Errorf("error unmarshaling %v: %v", key, err)
+	}
+
+	replenish := false
+	tb, ok := cfg.ReplenishmentStrategy.(*storagepb.Config_TimeBased)
+	if ok && now.Unix() >= bucket.LastReplenishMillisSinceEpoch/1000+tb.TimeBased.ReplenishIntervalSeconds {
+		bucket.Tokens += tb.TimeBased.TokensToReplenish
+		if bucket.Tokens > cfg.MaxTokens {
+			bucket.Tokens = cfg.MaxTokens
+		}
+		bucket.LastReplenishMillisSinceEpoch = now.UnixNano() / 1000000
+		replenish = true
+	}
+
+	if ok && add > 0 {
+		add = 0 // Do not replenish time-based quotas
+	}
+	before := bucket.Tokens
+	bucket.Tokens += add
+	if bucket.Tokens < 0 {
+		return 0, fmt.Errorf("insufficient tokens on %v (%v vs %v)", key, before, -add)
```

Comment:
We only fall into this condition if add is negative. If we don't negate it in the error message it'll be a confusing error.